### ISSUE TITLE
recent intel compiler issues

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#4878]
- [x] fix compiler detection: see #4877 
- [x] check compilation with 15, 16, and newer, fix problems
- [x] setup nightly regression tester and submit to cdash
- [x] fix tests
- [ ] make sure vectorization (including AVX2) now work correctly, see #4440 and #4978
- [ ] celebrate

**Comment by tamiko** [comment]
See also #4440

**Comment by drwells** [comment]
@masterleinad You merged a bunch of intel-related patches recently. Does this mean that we can close this, or at least check off some of these items?

**Comment by masterleinad** [comment]
I think its fair to say that compilation works right now.
Since I never had any real problem with vectorization and ICC, I am hesitant to check off that one.
The currently [failing tests](https://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=12075) (speficic to ICC18) are:

- bits/distorted_cells_04
- bits/distorted_cells_05
- fe/system_03
- fe/system_04
- matrix_free/shape_info
- matrix_free/thread_correctness

**Comment by kronbichler** [comment]
`thread_correctness` fails because some roundoff difference is `1e-6` for a float number, which is ok. `shape_info` is also uncritical because it fails for a Lagrange element on equidistant nodes for degree 17 which are very bad polynomials indeed. I relaxed that in #5226.

**Comment by kronbichler** [comment]
Regarding `thread_correctness`: We have a number of tests that are still sensitive to roundoff tolerances: Some tests use `float` numbers and show differences in the range of `1e-6` or `1e-5` which is OK but too large for `numdiff` where we assume absolute tolerance `1e-8` and relative tolerance `1e-6`. I checked on some tests of my machine here: 
https://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=11970

Most of the failures are due to such differences. I think we could work around the issue by a specific command
```
template <typename Number>
Number set_small_float_numbers_to_zero (const Number number, const double)
{
  return number;
}
float set_small_float_numbers_to_zero (const float number, const double tolerance)
{
  if (std::abs(number) < tolerance)
    return 0.f;
  else
    return number;
}
```
I suggest to add this to `tests.h` and use it in *selected circumstances* because with #4769 that removed the logstream manipulations in tests we cannot control roundoff differences on the order of `1e-6` - and we hardly want to add separate files. What do others think?

**Comment by tjhei** [comment]
> and we hardly want to add separate files

I agree. Alternative outputs are hard to maintain and should be a last resort.

in full_matrix_11.debug we fail because
> - DEAL::Difference: 0
> + DEAL::Difference: 7.62939e-06

I would just change the test to print "difference < 1e-5" if that condition is true.

**Comment by masterleinad** [comment]
I reported the failing tests `fe/system_03` and `fe_system_04` [here](https://software.intel.com/en-us/forums/intel-c-compiler/topic/746659).

**Comment by drwells** [comment]
I think that we fixed everything here as part of the 9.0 release. May we close this issue?

**Comment by masterleinad** [comment]
I am not aware of any specific problem that could be our fault.
